test/NuGet.Clients.Tests/NuGet.PackageManagement.UI.Test/Models/PackageItemListViewModelTests.cs

```diff
@@ -37,5 +54,424 @@ public void LocalSources_PackageReader_NotNull()
+                };
+
+                // Act
+                BitmapSource result = await GetFinalIconBitmapAsync(packageItemListViewModel);
+
+                // Assert
+                VerifyImageResult(result, packageItemListViewModel.BitmapStatus);
+                Assert.Equal(IconBitmapStatus.DefaultIconDueToDecodingError, packageItemListViewModel.BitmapStatus);
+            }
+        }
+
+        [Theory]
+        [MemberData(nameof(EmbeddedTestData))]
+        public void IsEmbeddedIconUri_Tests(Uri testUri, bool expectedResult)
+        {
+            var result = PackageItemListViewModel.IsEmbeddedIconUri(testUri);
+            Assert.Equal(expectedResult, result);
+        }
+
+        public static IEnumerable<object[]> EmbeddedTestData()
+        {
+            Uri baseUri = new Uri(@"C:\path\to\package");
+            var builder1 = new UriBuilder(baseUri)
+            {
+                Fragment = "    " // UriBuilder trims the string
+            };
+            var builder2 = new UriBuilder(baseUri)
+            {
+                Fragment = "icon.png"
+            };
+            var builder3 = new UriBuilder(baseUri)
+            {
+                Fragment = @"..\icon.png"
+            };
+            var builder4 = new UriBuilder(baseUri)
+            {
+                Fragment = string.Empty // implies that there's a Tag, but no value
+            };
+            var builder5 = new UriBuilder(baseUri)
+            {
+                Query = "aParam"
+            };
+            return new List<object[]>
+            {
+                new object[]{ builder1.Uri, false },
+                new object[]{ builder2.Uri, true },
+                new object[]{ builder3.Uri, true },
+                new object[]{ builder4.Uri, false },
+                new object[]{ builder5.Uri, false },
+                new object[]{ new Uri("https://sample.uri/"), false },
+                new object[]{ baseUri, false },
+                new object[]{ new Uri("https://another.uri/#"), false },
+                new object[]{ new Uri("https://complimentary.uri/#anchor"), false },
+                new object[]{ new Uri("https://complimentary.uri/?param"), false },
+                new object[]{ new Uri("relative/path", UriKind.Relative), false },
+            };
+        }
+
+        /// <summary>
+        /// Tests the final bitmap returned by the view model, by waiting for the BitmapStatus to be "complete".
+        /// </summary>
+        /// <param name="packageItemListViewModel"></param>
+        /// <returns></returns>
+        private static async Task<BitmapSource> GetFinalIconBitmapAsync(PackageItemListViewModel packageItemListViewModel)
+        {
+            BitmapSource result = packageItemListViewModel.IconBitmap;
+
+            while (!IconBitmapStatusUtility.GetIsCompleted(packageItemListViewModel.BitmapStatus))
+            {
+                await Task.Delay(250);
```

Comment:
Is there a possibility that this could hang a test?  Perhaps it should explicitly fail after `x` seconds.

Reply:
made it retry for 3 seconds. fixed in latest iteration.